### DISCUSSION TITLE
getStaticProps returns empty object at prod and dev (next v9.5.1)

### DISCUSSION THREAD

**Discussion by omar-dulaimi** [vercel/next.js#15993] [Help] [answered]
# Bug report

## Describe the bug

getStaticProps returns empty object at both dev and prod

## To Reproduce

```
import Header from './header';
import Head from './head';
import Footer from './footer';
import { useEffect } from 'react';
import Utils from '../util/utils';
import { hostName } from '../config/variables';

function Privacy({ pageTitle, metaTitle, description, metaDescription, url }) {
    console.log("Privacy -> pageTitle", pageTitle)
    useEffect(() => {
        Utils.handleFBPixelPageView('Privacy');
    }, []);

    return (
        <div>
            <Head
                isAmpPage={false}
                pageTitle={pageTitle}
                description={description}
                url={url}
                metaTitle={metaTitle}
                metaDescription={metaDescription}
            />
            <Header isAmp={false} />
            <div className="privacy-main-panel">

            Hello there!
            </div>

            <style jsx global>{`
                .privacy-main-panel {
                    width: 1000px;
                    display: flex;
                    flex-direction: column;
                    text-align: right;
                    font-family: inherit;
                    padding-top: 20px;
                    margin: 0 auto;
                }
                h1 {
                    padding: 0;
                    margin: 5px 0;
                    line-height: 1;
                    margin-bottom: 25px;
                }
                h2,
                h3 {
                    padding: 0;
                    margin: 5px 0;
                    line-height: 1;
                }
                p {
                    margin-bottom: 10px;
                }
                ul {
                    margin: 0;
                    margin-bottom: 20px;
                }
                @media only screen and (max-width: 1000px) {
                    .privacy-main-panel {
                        width: 100vw;
                        padding: 20px;
                    }
                }
            `}</style>
            <Footer isVisibleOnMobile={true} />
        </div>
    );
}

export async function getStaticProps() {
    return {
        props: {
            pageTitle: 'text1',
            metaTitle: 'text2',
            description: 'text3',
            metaDescription: 'text4',
            url: `${hostName}/privacy`,
        },
    };
}

export default Privacy;
```

## Expected behavior

Returned props should get passed to the component

## System information

- OS: Ubuntu 18.04
- Browser: Chrome Version 84.0.4147.105 (Official Build) (64-bit)
- Version of Next.js: 9.5.1
- Version of Node.js: 12.18.3

## Additional context

I dug through next-server.js and logged the related values: 
```
}renderToHTML -> pathname /privacy
result {
  components: {
    App: [Function: WrappedApp] {
      displayName: 'withRedux(f)',
      getInitialProps: [Function]
    },
    Document: [class MyDocument extends Document],
    Component: [Function: Privacy],
    buildManifest: {
      polyfillFiles: [Array],
      devFiles: [Array],
      ampDevFiles: [Array],
      lowPriorityFiles: [Array],
      pages: [Object],
      ampFirstPages: []
    },
    reactLoadableManifest: { './dev/noop': [Array] },
    pageConfig: {},
    getServerSideProps: undefined,
    getStaticProps: [AsyncFunction: getStaticProps],
    getStaticPaths: undefined
  },
  query: { _nextDataReq: undefined, amp: undefined }
}
this.renderOpts {
  poweredByHeader: true,
  canonicalBase: '',
  buildId: 'development',
  generateEtags: true,
  previewProps: {
    previewModeId: 'k',
    previewModeSigningKey: 'k',
    previewModeEncryptionKey: 'k'
  },
  customServer: true,
  ampOptimizerConfig: undefined,
  basePath: '',
  optimizeFonts: false,
  fontManifest: null,
  assetPrefix: '',
  dev: true,
  ErrorDebug: [Function: ReactDevOverlay],
  ampSkipValidation: false,
  ampValidator: [Function]
}
components.getStaticProps:  Promise {
  {
    props: {
      pageTitle: 'text1',
      metaTitle: 'text2',
      description: 'text3',
      metaDescription: 'text4',
      url: 'http://localhost:3000/privacy'
    }
  }
}
query:  [Object: null prototype] {}
params:  null
all:  { _nextDataReq: undefined, amp: undefined }
Privacy -> pageTitle undefined

```
SO link: https://stackoverflow.com/questions/63312375/getstaticprops-returns-empty-object-at-prod-and-dev-next-v9-5-1

**Comment by desaijay315** [comment]
Can you please check if the pageProps is passed correctly to the  custom component

export default ({ Component, pageProps }) => (
  <Component {...pageProps} />
)

**Comment by omar-dulaimi** [comment]
Hey @desaijay315 
If by custom component you meant the _app, then yeah it's correctly passed:
```
        const { Component, pageProps, store } = this.props;
        return (
            <Provider store={store}>
                <Component {...pageProps} />
                <MobileBarController />
            </Provider>
        );
```

**Reply by desaijay315** [reply]
Can you share the GitHub repo -----

**Reply by omar-dulaimi** [reply]
Unfortunately I'm not allowed to do that

**Comment by omar-dulaimi** [comment]
Why did this issue get converted to a discussion?
The original discussion received no attention since yesterday and its clearly a real issue and not just a typo or something!

**Reply by omar-dulaimi** [reply]
Hope the ones that made the decision help with an answer soon instead of keeping it unsolved like this

**Comment by omar-dulaimi** [comment]
The problem seems to be here:

```
      static async getInitialProps({ Component, ctx }) {
        const pageProps = Component.getInitialProps
            ? await Component.getInitialProps(ctx)
            : {};
        return { pageProps };
    }
```
Component.getInitialProps is undefined so pageProps returns an empty object.

Does this mean that getStaticProps doesn't actually pass props to getInitialProps in the custom app?

**Reply by desaijay315** [reply]
Since you are using getStaticProps,  _app doesn't have getStaticProps, you can still use getInitialProps for other pages. This is the default behaviour, not a problem

**Reply by omar-dulaimi** [reply]
But what about the pages that use getStaticProps? Aren't they supposed to get their props here?

**Comment by omar-dulaimi** [comment]
Both appProps and pageProps are empty:
```
 const pageProps = AppContext.Component.getInitialProps ? 
     await AppContext.Component.getInitialProps(AppContext.ctx)
     : {};

 console.log('MyApp -> getInitialProps -> pageProps', pageProps);

 const appProps = App.getInitialProps ? await App.getInitialProps(AppContext): {};
 console.log('MyApp -> getInitialProps -> appProps', appProps);
```

**Reply by desaijay315** [reply]
**Already getting the results!**

**Can you provide more information on below components -**. 

Also on useEffect - 

```
import Header from './header';
import Head from './head';
import Footer from './footer';
import Utils from '../util/utils';
import { hostName } from '../config/variables';
```

![image](https://user-images.githubusercontent.com/9349985/89727050-13e83000-da3f-11ea-8d7f-4583ce4845c8.png)

**Reply by omar-dulaimi** [reply]
This is the header:
```
import { primaryColor, darkColor } from '../util/style';
import Link from 'next/link';
import Logo from '../components/icons/logo';
import { connect } from 'react-redux';
import Avatar from '../components/avatar';

function Header(props) {
    let { user } = props;
    return (
        <div className="nav-panel">
            <div className="header-main-panel">
                <div style={{ display: 'flex', float: 'left' }}>
               
                    <Link href={`/new-list`} prefetch={false}>
                        <a className="add-new-btn">add new listing</a>
                    </Link>
                </div>
            </div>

            <style jsx global>{`
                .nav-panel {
                    background-color: #f8f9fa;
                    height: 68px;
                }
                .header-main-panel {
                    width: 1000px;
                    margin: 0 auto;
                    display: flex;
                    justify-content: space-between;
                }

                .add-new-btn {
                    background: ${primaryColor};
                    margin-left: 20px;
                    border-radius: 5px;
                    padding: 0 10px;
                    text-decoration: none;
                    color: #fff;
                    line-height: 40px;
                    margin-top: 12px;
                    height: 40px;
                    font-weight: bold;
                }
                .sign-in-btn {
                    padding: 0 10px;
                    text-decoration: none;
                    color: ${darkColor};
                    line-height: 60px;
                    float: left;
                }
                .user-image-panel {
                    position: absolute;
                    top: 50%;
                    left: 50%;
                    transform: translate(-50%, -50%);
                    border-radius: 50%;
                    max-width: 100%;
                    max-height: 100%;
                    overflow: hidden;
                }
                .user-image {
                    width: 80px;
                    height: auto;
                }
                .banner-user-image {
                    width: 32px;
                    border-radius: 50%;
                }
                .author-user-image {
                }
                @media only screen and (max-width: 1000px) {
                    .nav-panel {
                        display: none;
                    }
                }
            `}</style>
        </div>
    );
}

const mapStateToProps = (state) => ({
    user: state.userReducer.user,
});

export default connect(mapStateToProps)(Header);

```

**Reply by omar-dulaimi** [reply]
The Head just sets meta tags according to page type; amp or not, etc.

**Reply by omar-dulaimi** [reply]
The Footer is just a basic page with static info.

**Reply by omar-dulaimi** [reply]
Utils is a helper class with many static methods.

**Reply by omar-dulaimi** [reply]
And this import { hostName } from '../config/variables';
just reads an env var from env, process.env.hostName

**Reply by omar-dulaimi** [reply]
Do you have a custom app as well?

**Comment by omar-dulaimi** [comment]
useEffect just handles page views: 
```
    static handleFBPixelPageView(page) {
        const md = new MobileDetect(navigator.userAgent);
        const deviceOs = md.os();
        ReactPixel.fbq('track', 'PageView');
        ReactPixel.fbq('track', page, { os: deviceOs ? deviceOs : 'desktop' });
    }
```

**Comment by omar-dulaimi** [comment]
Hello @desaijay315 
Any luck?
If nobody is capable of solving it, then maybe it was a real issue? Maybe it shouldn't have been converted to a discussion ?

**Reply by desaijay315** [reply]
Can you create the GitHub repo with code temporarily to reproduce the issue?

**Reply by omar-dulaimi** [reply]
Okay, I managed to create a repro here: https://github.com/omar-dulaimi/nextjs-no-staticprops-repro
please check it out

**Comment by omar-dulaimi** [comment, marked as answer]
@desaijay315 The issue is related to next-redux-wrapper library. It introduced new changes after Nextjs's latest features, their way of handling getStaticProps has changed, and since I only upgraded NextJs this issue occurred.

The docs should mention this, and next-redux-wrapper should display a warning.

Anyways, thanks for your time!

**Comment by desaijay315** [comment]
That's great that you found the issue....Please mention the issue and also
let the Devs at Next.js know about it...they will upgrade the docs

**Comment by wparam** [comment]
I have this issue too, is this issue solved? Or we're waiting for next-redux-wrapper team's fix?

**Comment by usman786-ctrl** [comment]
facing the same issue and not getting the solution

**Comment by omar-dulaimi** [comment]
@wparam @usman786-ctrl 
In my case this occurred when I updated Nextjs without updating the syntax of next-redux-wrapper inside my app.
If you do use that library, make sure to go over their new syntax on their github repo. https://github.com/kirill-konshin/next-redux-wrapper

As you can see, there's a warning on top of the page saying this:
`The current version of this library only works with Next.js 9.3 and newer. If you are required to use Next.js 6-9 you can use version 3-5 of this library, see branches. Otherwise, consider upgrading Next.js.`

**Reply by TotallWAR** [reply]
Man, you saved DAYS of my work! Thank you very much!!!

**Reply by omar-dulaimi** [reply]
You're welcome, glad I could help

**Comment by brunograna** [comment]
How did you solved this issue? Facing the same problem here

**Reply by omar-dulaimi** [reply]
> In my case this occurred when I updated Nextjs without updating the syntax of next-redux-wrapper inside my app.
> If you do use that library, make sure to go over their new syntax on their github repo. https://github.com/kirill-konshin/next-redux-wrapper
> 
> As you can see, there's a warning on top of the page saying this:
> The current version of this library only works with Next.js 9.3 and newer. If you are required to use Next.js 6-9 you can use version 3-5 of this library, see branches. Otherwise, consider upgrading Next.js.


So yeah, just upgrade next-redux-wrapper to the latest version and change how you use nextjs methods like getServerSideProps, getStaticProps, etc to their equivalent syntax.

**Reply by subgero** [reply]
It would be nice if you share your code or elaborate your answer, this will help to other dev's when they face this.

**Comment by lovanya** [comment]
I created the same issue in next.js but they moved it to discussions. Now finally can solve the problem. Thanks a lot @omar-dulaimi .

**Comment by Shorifpatwary** [comment]
Maybe you are calling the **getStaticProps** or **getServerSideProps** or **getStaticPaths** function from outside of pages or app directory (v-13). If you do that next js will send you an empty props object. 
_**Next.js does not throw any error message to catch this error properly.**_
